apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -0,0 +1,2151 @@
+use serde_json_bytes::Map as JSONMap;
+use serde_json_bytes::Value as JSON;
+
+use super::helpers::json_type_name;
+use super::immutable::InputPath;
+use super::lit_expr::LitExpr;
+use super::ApplyTo;
+use super::ApplyToError;
+use super::MethodArgs;
+use super::PathList;
+use super::VarsWithPathsMap;
+
+type ArrowMethod = fn(
+    // Method name
+    &str,
+    // Arguments passed to this method
+    &Option<MethodArgs>,
+    // The JSON input value (data)
+    &JSON,
+    // The variables
+    &VarsWithPathsMap,
+    // The input_path (may contain integers)
+    &InputPath<JSON>,
+    // The rest of the PathList
+    &PathList,
+    // Errors
+    &mut IndexSet<ApplyToError>,
+) -> Option<JSON>;
+
+lazy_static! {
+    // This set controls which ->methods are exposed for use in connector
+    // schemas. Non-public methods are still implemented and tested, but will
+    // not be returned from lookup_arrow_method outside of tests.
+    static ref PUBLIC_ARROW_METHODS: IndexSet<&'static str> = {
+        let mut public_methods = IndexSet::default();
+
+        public_methods.insert("echo");
+        // public_methods.insert("typeof");
+        public_methods.insert("map");
+        // public_methods.insert("eq");
+        public_methods.insert("match");
+        // public_methods.insert("matchIf");
+        // public_methods.insert("match_if");
+        // public_methods.insert("add");
+        // public_methods.insert("sub");
+        // public_methods.insert("mul");
+        // public_methods.insert("div");
+        // public_methods.insert("mod");
+        public_methods.insert("first");
+        public_methods.insert("last");
+        public_methods.insert("slice");
+        public_methods.insert("size");
+        // public_methods.insert("has");
+        // public_methods.insert("get");
+        // public_methods.insert("keys");
+        // public_methods.insert("values");
+        public_methods.insert("entries");
+        // public_methods.insert("not");
+        // public_methods.insert("or");
+        // public_methods.insert("and");
+
+        public_methods
+    };
+
+    // This map registers all the built-in ->methods that are currently
+    // implemented, even the non-public ones that are not included in the
+    // PUBLIC_ARROW_METHODS set.
+    static ref ARROW_METHODS: IndexMap<String, ArrowMethod> = {
+        let mut methods = IndexMap::<String, ArrowMethod>::default();
+
+        // This built-in method returns its first input argument as-is, ignoring
+        // the input data. Useful for embedding literal values, as in
+        // $->echo("give me this string").
+        methods.insert("echo".to_string(), echo_method);
+
+        // Returns the type of the data as a string, e.g. "object", "array",
+        // "string", "number", "boolean", or "null". Note that `typeof null` is
+        // "object" in JavaScript but "null" for our purposes.
+        methods.insert("typeof".to_string(), typeof_method);
```

Comment:
Let's make sure that this gets documented to users as it diverges from what they might expect.

Reply:
On top of `typeof null == "object"`, `typeof [1,2,3]` also returns `"object"`, forcing developers to test `Array.isArray` separately, so that's another (helpful!) deviation from JS behavior.

Definitely worth documenting (as with all public `->` methods), but also a defensible/welcome improvement over the ambiguities of `typeof` in JS.